src/guest_memory.rs

```diff
@@ -309,6 +304,134 @@ pub trait GuestMemory {
             std::cmp::max,
         )
     }
+}
+
+/// GuestMemorySnapshottable provides a way to retrieve a GuestMemory object.
+/// The vm-memory crate already provides trivial implementation for
+/// instances of GuestMemory (more precisely, for references to them
+/// or reference-counted GuestMemory objects), but the trait can also
+/// be implemented by a GuestMemoryMut object in order to obtain a
+/// consistent snapshot of the memory map.
+///
+/// In order to support generic mutable memory maps, devices (or other things
+/// that access memory) should store the memory as a GuestMemorySnapshottable<M>.
+/// In the following example, dummy structs that implement the GuestMemoryBackend trait
+/// describe how the VMM is storing the memory map.  We could also declare VirtioDevice
+/// as `<M: GuestMemory, S: GuestMemorySnapshottable<M>`, but this would lead
+/// to repetitive types such as `VirtioDevice<GuestMemoryMmap, &GuestMemoryMmap>`.
+///
+/// ```
+/// # #[cfg(feature = "backend-mmap")]
+/// # use vm_memory::GuestMemoryMmap;
+/// # #[cfg(feature = "backend-atomic")]
+/// # use vm_memory::GuestMemoryAtomic;
+/// # use vm_memory::{GuestAddress, GuestMemory, GuestMemorySnapshottable};
+///
+/// # // Perhaps this could be added to vm-memory itself?
```

Comment:
Definitely it should be included in vm-memory, because it's may be the main interfaces used by consumers such as vm-virtio.

Reply:
Ok, done.